agent/cache/watch.go

```diff
@@ -106,148 +106,192 @@ func (c *Cache) notifyBlockingQuery(ctx context.Context, r getOptions, correlati
+	// Check context hasn't been canceled
+	if ctx.Err() != nil {
+		return
+	}
 
-			// Update index for next request
-			index = meta.Index
-		}
+	// Check the index of the value returned in the cache entry to be sure it
+	// changed
+	if index == 0 || index < meta.Index {
+		cb(ctx, UpdateEvent{correlationID, res, meta, err})
 
-		var wait time.Duration
-		// Handle errors with backoff. Badly behaved blocking calls that returned
-		// a zero index are considered as failures since we need to not get stuck
-		// in a busy loop.
-		if err == nil && meta.Index > 0 {
-			failures = 0
-		} else {
-			failures++
-			wait = backOffWait(failures)
-
-			c.options.Logger.
-				With("error", err).
-				With("cache-type", r.TypeEntry.Name).
-				With("index", index).
-				Warn("handling error in Cache.Notify")
-		}
+		// Update index for next request
+		index = meta.Index
+	}
 
-		if wait > 0 {
-			select {
-			case <-time.After(wait):
-			case <-ctx.Done():
-				return
-			}
-		}
-		// Sanity check we always request blocking on second pass
-		if err == nil && index < 1 {
-			index = 1
+	var wait time.Duration
+	// Handle errors with backoff. Badly behaved blocking calls that returned
+	// a zero index are considered as failures since we need to not get stuck
+	// in a busy loop.
+	if err == nil && meta.Index > 0 {
+		failures = 0
+	} else {
+		failures++
+		wait = backOffWait(failures)
+
+		c.options.Logger.
+			With("error", err).
+			With("cache-type", r.TypeEntry.Name).
+			With("index", index).
+			Warn("handling error in Cache.Notify")
+	}
+
+	if wait > 0 {
+		select {
+		case <-time.After(wait):
+		case <-ctx.Done():
+			return
 		}
 	}
+	// Sanity check we always request blocking on second pass
+	if err == nil && index < 1 {
+		index = 1
+	}
+
+	return
 }
 
 func (c *Cache) notifyPollingQuery(ctx context.Context, r getOptions, correlationID string, cb Callback) {
-	index := uint64(0)
-	failures := uint(0)
-
-	var lastValue interface{} = nil
+	var (
+		index         = uint64(0)
+		failures      = uint(0)
+		lastValue any = nil
+	)
 
 	for {
 		// Check context hasn't been canceled
-		if ctx.Err() != nil {
+		select {
+		case <-ctx.Done():
 			return
+		default:
 		}
 
-		// Make the request
-		r.Info.MinIndex = index
-		res, meta, err := c.getWithIndex(ctx, r)
+		index, failures, lastValue = c.notifyPollingQueryOnce(ctx, r, correlationID, cb, index, failures, lastValue)
+	}
+}
+func (c *Cache) notifyPollingQueryOnce(
+	ctx context.Context,
+	r getOptions,
+	correlationID string,
+	cb Callback,
+	currIndex uint64,
+	currFailures uint,
+	currValue any,
+) (index uint64, failures uint, lastValue any) {
+	// Seed response data from last execution.
+	index = currIndex
+	failures = currFailures
+	lastValue = currValue
 
-		// Check context hasn't been canceled
-		if ctx.Err() != nil {
-			return
-		}
+	// Ensure that each invocation terminates instead of stacking.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
I think this would work here because the wait is done at this level so by the time cancel is called we really want to cancel the context because we waited enough and it was not done.